### PR TITLE
Srravich/feature/2039 restart conversation

### PR DESCRIPTION
**Pull request by srinaath**
Fixes https://github.com/microsoft/BotFramework-Emulator/issues/2039

This PR adds the capability to replay activities in a conversation. It allows multiple conversations being replayed at the same time.

Screen Shot 2020-02-28 at 5 13 10 PM

Screen Shot 2020-02-28 at 5 05 56 PM

FinalReplay

### CODE REVIEW

File: packages/app/client/webpack.config.js [hunk 1]
```diff
@@ -107,7 +106,7 @@ const defaultConfig = {
     ],
   },
 
-  devtool: 'source-map',
+  devtool: 'eval-source-map',
```
Comment (tonyanziano) [inline packages/app/client/webpack.config.js]: Does this fix the Web Chat source maps and preserve the Emulator source maps? I tried to change it locally from `devtool: sourcemap` to using the `SourceMapDevTool plugin` from web chat and it managed to fix the Web Chat source maps, but downgraded the quality of the Emulator source maps to point to the transpiled code.

Comment (srinaath) [reply packages/app/client/webpack.config.js]: It does not Tony. So it still shows the transpiled code for webchat and debuggable code for emulator. So I was facing an issue with step through debugging on Chrome Dev tools failing when i use source-map. The moment I use eval-source-maps it does not break or Chrome dev tools does not seem to break when i step though and debug. Have u faced this issue before?

File: packages/app/client/src/state/sagas/chatSagas.ts [hunk 2]
```diff
@@ -226,110 +311,176 @@ export class ChatSagas {
...
+            break;
+          }
+        }
+      } catch (err) {
+        wcEventChannel.close();
+        // Restart the channel if error occurs
+        ChatSagas.wcActivityChannel = createWebchatActivityChannel();
       } finally {
-        yield put(updatePendingSpeechTokenRetrieval(documentId, false));
+        yield fork(ChatSagas.handleReplayIfRequired, { documentId, action, dispatch, meta });
```
Comment (tonyanziano) [inline packages/app/client/src/state/sagas/chatSagas.ts]: why does `dispatch` need to be passed in if we have access to `put` ?

File: packages/app/client/src/state/sagas/chatSagas.ts [hunk 3]
```diff
@@ -495,7 +678,10 @@ export class ChatSagas {
     const secret = encode(JSON.stringify(options));
     const res: Response = yield fetch(`${serverUrl}/emulator/ws/port`);
     if (!res.ok) {
-      yield* throwErrorFromResponse('Error occurred while retrieving the web socket port', res);
+      throw new Error(
+        `Error occurred while retrieving the WebSocket server port: ${res.status}: ${res.statusText ||
+          'No status text'}`
+      );
```
Comment (tonyanziano) [inline packages/app/client/src/state/sagas/chatSagas.ts]: change back to `throwErrorFromResponse`?

File: packages/app/shared/src/state/actions/chatActions.ts [hunk 4]
```diff
@@ -332,14 +354,51 @@ export function showContextMenuForActivity(activity: Partial<Activity>): ChatAct
 export function restartConversation(
   documentId: string,
   requireNewConversationId: boolean = false,
-  requireNewUserId: boolean = false
+  requireNewUserId: boolean = false,
+  activity: Activity = undefined,
+  createObjectUrl: Function = undefined
```
[the remark this answers is not in the document]
Comment (srinaath) [reply packages/app/shared/src/state/actions/chatActions.ts]: What do u think?

File: packages/app/client/src/state/sagas/chatSagas.ts (hunk 2 above)
Comment (tonyanziano) [reply packages/app/client/src/state/sagas/chatSagas.ts]: Ahhh, I didn't catch that. Sounds good to me 👍

File: packages/app/client/webpack.config.js (hunk 1 above)
Comment (srinaath) [reply packages/app/client/webpack.config.js]: I have reverted it to use just source-map. Seems like we can come up with a good solution for R9 that enables source maps for emulator as well as enable it for our shared libs (webchat, shared/app, shared/sdk)

File: packages/sdk/ui-react/src/widget/splitButton/splitButton.tsx [hunk 5]
```diff
@@ -140,28 +141,33 @@ export class SplitButton extends React.Component<SplitButtonProps, SplitButtonSt
...
-    if (onClick && options.length) {
-      onClick(options[0]);
+  private onClickDefault = (): void => {
+    const { onDefaultButtonClick, options = [] } = this.props;
+    if (onDefaultButtonClick && options.length) {
+      if (!this.state.selected) {
+        onDefaultButtonClick(options[0]);
+      } else {
+        onDefaultButtonClick(options[this.state.selected]);
+      }
```
[the remark this answers is not in the document]
Comment (srinaath) [reply packages/sdk/ui-react/src/widget/splitButton/splitButton.tsx]: We resolved this in a separate thread. Basically, the decision is to show the selected option on choosing an option

File: packages/app/client/src/ui/editor/emulator/parts/chat/outerActivityWrapper.tsx [hunk 6]
```diff
@@ -42,29 +44,57 @@ export interface OuterActivityWrapperProps {
...
+  documentId: string;
   onContextMenu?: (event: React.MouseEvent<HTMLElement>) => void;
   onItemRendererClick?: (event: React.MouseEvent<HTMLElement>) => void;
   onItemRendererKeyDown?: (event: React.KeyboardEvent<HTMLElement>) => void;
+  onRestartConversationFromActivityClick?: (
+    documentId: string,
+    activity: Activity,
+    restartOption: RestartConversationOptions
+  ) => void;
+  currentRestartConversationOption: RestartConversationOptions;
```
[the remark this answers is not in the document]
Comment (srinaath) [reply packages/app/client/src/ui/editor/emulator/parts/chat/outerActivityWrapper.tsx]: Confirmed. This is not happening. We are good to go with these changes

**Comment by srinaath** [comment]
> Looks good! Great job 😃

thanks Tony

**Pull request merged**